base/scopedvalues.jl

```diff
@@ -0,0 +1,190 @@
+
+function Base.show(io::IO, scope::Scope)
+    print(io, Scope, "(")
+    first = true
+    for (key, value) in scope.values
+        if first
+            first = false
+        else
+            print(io, ", ")
+        end
+        print(io, typeof(key), "@")
+        show(io, Base.objectid(key))
+        print(io, " => ")
+        show(IOContext(io, :typeinfo => eltype(key)), value)
+    end
+    print(io, ")")
+end
+
+struct NoValue end
+const novalue = NoValue()
+
+function get(val::ScopedValue{T}) where {T}
+    # Inline current_scope to avoid doing the type assertion twice.
+    scope = current_task().scope
+    if scope === nothing
+        isassigned(val) && return Some(val.default)
+        return nothing
+    end
+    scope = scope::Scope
+    if isassigned(val)
+        return Some(Base.get(scope.values, val, val.default)::T)
+    else
+        v = Base.get(scope.values, val, novalue)
+        v === novalue || return Some(v::T)
+    end
+    return nothing
+end
+
+function Base.getindex(val::ScopedValue{T})::T where T
+    maybe = get(val)
+    maybe === nothing && throw(KeyError(val))
+    return something(maybe)::T
+end
+
+function Base.show(io::IO, val::ScopedValue)
+    print(io, ScopedValue)
+    print(io, '{', eltype(val), '}')
+    print(io, '(')
+    v = get(val)
+    if v === nothing
+        print(io, "undefined")
+    else
+        show(IOContext(io, :typeinfo => eltype(val)), something(v))
+    end
+    print(io, ')')
+end
+
+"""
+    with(f, (var::ScopedValue{T} => val::T)...)
+
+Execute `f` in a new scope with `var` set to `val`.
+"""
+function with(f, pair::Pair{<:ScopedValue}, rest::Pair{<:ScopedValue}...)
+    @nospecialize
+    ct = Base.current_task()
+    current_scope = ct.scope::Union{Nothing, Scope}
+    scope = Scope(current_scope, pair, rest...)
+    ct.scope = scope
+    try
+        return f()
+    finally
+        ct.scope = current_scope
+    end
+end
+
+with(@nospecialize(f)) = f()
+
+"""
+    @with vars... expr
+
+Macro version of `with(f, vars...)` but with `expr` instead of `f` function.
+This is similar to using [`with`](@ref) with a `do` block, but avoids creating
+a closure.
+"""
+macro with(exprs...)
+    if length(exprs) > 1
+        ex = last(exprs)
+        exprs = exprs[1:end-1]
+    elseif length(exprs) == 1
+        ex = only(exprs)
+        exprs = ()
+    else
+        error("@with expects at least one argument")
+    end
+    for expr in exprs
+        if expr.head !== :call || first(expr.args) !== :(=>)
+            error("@with expects arguments of the form `A => 2` got $expr")
+        end
+    end
+    exprs = map(esc, exprs)
+    ct = gensym(:ct)
+    current_scope = gensym(:current_scope)
+    body = Expr(:tryfinally, esc(ex), :($(ct).scope = $(current_scope)))
+    quote
+        $(ct) = $(Base.current_task)()
+        $(current_scope) = $(ct).scope::$(Union{Nothing, Scope})
+        $(ct).scope = $(Scope)($(current_scope), $(exprs...))
+        $body
+    end
```

Comment:
I think this should be sufficient (`task` and `current_scope` should be gensymed automatically)
```suggestion
    quote
        task = current_task()
        current_scope = task.scope::Union{Nothing, Scope}
        task.scope = Scope(current_scope, $(map(esc,exprs)...))
        $(Expr(:tryfinally, esc(ex), :(task.scope = current_scope)))
    end
```

Reply:
Names from module e.g Scope and Base.current_task need to be interpolated.

Reply:
Names are qualified by the current module unless you `esc`, e.g.:
```julia
julia> module TestScope
       macro with(exprs...)
           if length(exprs) > 1
               ex = last(exprs)
               exprs = exprs[1:end-1]
           elseif length(exprs) == 1
               ex = only(exprs)
               exprs = ()
           else
               error("@with expects at least one argument")
           end
           for expr in exprs
               if expr.head !== :call || first(expr.args) !== :(=>)
                   error("@with expects arguments of the form `A => 2` got $expr")
               end
           end

           quote
               task = Base.current_task()
               current_scope = task.scope::Union{Nothing, Scope}
               task.scope = Scope(current_scope, $(map(esc,exprs)...))
               $(Expr(:tryfinally, esc(ex), :(task.scope = current_scope)))
           end
       end
       end
Main.TestScope

julia> @macroexpand TestScope.@with foo=>2 bar(x)
quote
    #= REPL[1]:19 =#
    var"#3#task" = (Main.TestScope.Base).current_task()
    #= REPL[1]:20 =#
    var"#4#current_scope" = (var"#3#task").scope::Main.TestScope.Union{Main.TestScope.Nothing, Main.TestScope.Scope}
    #= REPL[1]:21 =#
    (var"#3#task").scope = Main.TestScope.Scope(var"#4#current_scope", foo => 2)
    #= REPL[1]:22 =#
    $(Expr(:tryfinally, :(bar(x)), :((var"#3#task").scope = var"#4#current_scope")))
end
```